### DISCUSSION TITLE
Async `session.refresh()` "refreshes away" related objects

### DISCUSSION THREAD

**Discussion by ljiatu** [sqlalchemy/sqlalchemy#6243] [Usage Questions] [answered]
In the following example, `B` objects seem to be marked as stale and their attributes no longer accessible once `A` is refreshed in the session. This could be the expected behavior, and I never noticed when using synchronous sessions since implicit attribute load was OK, but want to double-check.
```py
import asyncio

from sqlalchemy import Column, ForeignKey, Integer
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import backref, relationship, sessionmaker

Base = declarative_base()


class B(Base):
    __tablename__ = "bs"

    id = Column(Integer, primary_key=True)
    a_id = Column(
        Integer,
        ForeignKey("as.id", onupdate="CASCADE", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )


class A(Base):
    __tablename__ = "as"

    id = Column(Integer, primary_key=True)
    bs = relationship(
        "B",
        backref=backref("a", lazy="raise"),
        cascade="all, delete-orphan",
        passive_deletes=True,
        lazy="raise",
    )


async def async_main():
    engine = create_async_engine(
        "postgresql+asyncpg://localhost/platform_dev",
        echo=True,
    )

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)

    session_maker = sessionmaker(
        engine, class_=AsyncSession, expire_on_commit=False
    )

    obj_a = A(id=1)
    obj_b1 = B(id=1, a=obj_a)
    obj_b2 = B(id=2, a=obj_a)
    async with session_maker() as session:
        async with session.begin():
            session.add(obj_a)
            session.add(obj_b1)
            session.add(obj_b2)

        print("-" * 80)
        print(obj_a.id)  # works fine
        print("-" * 80)
        await session.refresh(obj_a)
        print("*" * 80)
        print(obj_b1.id)  # throws exception because `obj_b1` appears to have been marked as stale
        print("*" * 80)

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)


asyncio.run(async_main())
```

The exception being thrown is `sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place? (Background on this error at: http://sqlalche.me/e/14/xd2s)`, as usually seen when implicit IO on attribute access is attempted.

**Comment by zzzeek** [comment, marked as answer]
hi -

yeah, this is expected, though certainly nothing anyone has ever noticed before we had asyncio :) .     

refresh() works by first expiring the object, then reloading it.    When it expires obj_a, the "expiration" object [cascades](https://docs.sqlalchemy.org/en/14/orm/cascades.html#refresh-expire) down to the related objects, due to the linked "refresh-expire" cascade, which in your above mapping, is set up because you indicated "all" in your cascade setting :) .

So....after that, the related object doesn't get refreshed in that step because there's no eager loading set up from A->B..which is normal here.

So.....here what you want to set up:

```
    bs = relationship(
        "B",
        backref=backref("a", lazy="raise"),
        cascade="save-update, merge, delete, delete-orphan",
        passive_deletes=True,
        lazy="raise",
    )

```

where I took out the "all", and replaced with the more specific "save-update, merge" cascades that's usually what people need.    This is worth adding to that doc section in asyncio.

**Reply by zzzeek** [reply]
I've added notes on this in 7ea0d10c67b652895442c3a103e0813769309b9c

**Reply by ljiatu** [reply]
Ah that makes a lot of sense. Thanks for the explanation! What's a scenario where `refresh-expire` could be useful?

**Reply by zzzeek** [reply]
well if you want to expire A it seems like often you'd want to expire all the things related to it.

Me personally, I dont really use refresh or expire at all unless there's some very special thing going on, i just work in one transaction and that's it.
